src/python/pants/backend/experimental/python/lockfile.py

```diff
@@ -124,7 +147,16 @@ async def generate_lockfile(
             output_files=(req.dest,),
         ),
     )
-    return PythonLockfile(result.output_digest, req.dest)
+
+    lockfile_contents = next(
+        iter(await Get(DigestContents, Digest, generated_lockfile.output_digest))
+    )
+    content_with_header = b"%b\n%b" % (lockfile_metadata(req.hex_digest), lockfile_contents.content)
+    complete_lockfile = await Get(
+        Digest, CreateDigest([FileContent(req.dest, content_with_header)])
+    )
```

Comment:
It'd be great to test this set of functionality: prepending a header to the lockfile. 

I don't think you need to necessarily use `RuleRunner` to test the `CreateDigest` and `DigestContents` part: those already have ample tests. More important I think is what goes in between. `RuleRunner` can be tricky also, makes the tests more complex.

I do think the test should ensure the `b"%b\n%b" % (lockfile_metadata(req.hex_digest), lockfile_contents.content)` line works correctly. If you end up not using `RuleRunner`, you'd want your normal Python function to handle that and move it out of the rule.

If you do use `RuleRunner`, it would be helpful to extract a helper rule that only handles the updating of the lockfile after it was generated. But that means new boilerpalte like defining a Request type and result type. I think better is to stick with what you have, but update your normal Python function as proposed and use a normal Pytest test for it. Wdyt?

Reply:
Done!